test/cpp/jit/test_lite_interpreter.cpp

```diff
@@ -609,14 +610,94 @@ TEST(LiteInterpreterTest, TwoSubmodulesModuleInfo) {
   AT_ASSERT(module_debug_info_set == expected_result);
 }
 
+TEST(LiteInterpreterTest, LoadAndRunByteCodeModel) {
```

Comment:
> My suggestion is to do something like this:
> 
> ```
> // A generic contract for backport logic to the previous bytecode version.
> // Args:
> // * PyTorchStreamReader refers to the model in the from N bytecode version.
> // * PyTorchStreamWriter refers to the model backported to the previous N-1 bytecode version.
> // Returns true if successful, false otherwise.
> using BytecodeBackportFunction =
>     std::function<bool(PyTorchStreamReader, PyTorchStreamWriter)>;
> 
> 
> // Registry of backport functions.
> void registerBytecodeBackportFunction(
> 	const int64_t from_version,
> 	const BytecodeBackportFunction& backport_function);
> 
> std::unordered_map<int64_t, BytecodeBackportFunction>&
> bytecodeBackportFunctions() {
>   static std::unordered_map<int64_t, BytecodeBackportFunction>
>       backport_functions;
>   return backport_functions;
> }
> 
> bool hasBytecodeBackportFunction(const int64_t from_version) {
>   return bytecodeBackportFunctions().count(from_version);
> }
> 
> void registerBytecodeBackportFunction(
>     const int64_t from_version,
> 	const BytecodeBackportFunction& backport_function) {
>   TORCH_CHECK(
>       !detail::hasBytecodeBackportFunction(from_version),
>       "Backporting from version ",
>       from_version,
>       " is already registered.");
>   bytecodeBackportFunctions()[from_version] = backport_function;
> }
> 
> // Then, the public backport API would look like:
> 
> bool _backport_to_bytecode_version(..., int64_t to_version) {
>   if (!hasBytecodeBackportFunction(to_version + 1)) return false;
>   int64_t bytecode_version = // get model bytecode version
>   // Assuming you get your PyTorchStreamReader, PyTorchStreamWriter
>   // Plus relevant checks about to_version and current bytecode version.
>   while (bytecode_version > to_version) {
>     if (!hasBytecodeBackportFunction(bytecode_version)) return false;
>     bytecodeBackportFunctions()[bytecode_version--](...);
>   }
>   return true;
> }
> ```

That's an awesome design! It's great to provide flexibility of each n to n-1 conversion, and also makes the tests easier to create! 

For implementation details, one option is to combine PyTorchStreamReader and PyTorchStreamWriter to a PyTorchStreamConverter. In such case the interface would be directly model-in, model-out, and easier to be piped.

Reply:
Actually, I wrote this before I proposed the code.
With the proposed interface, then it should be even easier: a person doing a version bump, and implementing the backport function, should only need to provide a model file in the new version, and the same test suite should be able to run by executing the backport API one by one towards the past versions and verify all the generic backport checks pass.

@iseeyuan

Reply:
backport itself is the converter interface. Basically the flow currently is

> model_in (stream, file..) -> reader -> backport -> writer -> model_out (stream, file...)

We can create a wrapper (mark with brackets) and call it converter

> model (stream, file..) -> [ reader -> backport -> writer ] -> model (stream, file...)